### PR TITLE
install / uninstall - Datadog Puppet Module w/ Windows support

### PR DESCRIPTION
**Pull request by jaycdave88**
This PR is created for a client request's to have official support for Puppet with their Windows environments.

### CODE REVIEW

**Comment by albertvaka** [comment]
Thanks for the patch @jaycdave88! Could you please fill in the PR description with some context/motivation for the change? Thanks 🙏

File: manifests/windows/agent6.pp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  String $msi_destination = $datadog_agent::params::win_msi_location,
+  String $api_key = $datadog_agent::api_key,
+  String $hostname = $datadog_agent::host,
+  String $service_name = $datadog_agent::service_name_win,
+  Array  $tags = $datadog_agent::tags,
+  Boolean $service_enable = true,
+  Boolean $should_install = true
+) inherits datadog_agent::params {
+
+	if $should_install {
```
Comment (albertvaka) [inline manifests/windows/agent6.pp]: Mixed tabs and spaces here and in the matching else below.

**Comment by jaycdave88** [comment]
> Thanks for the patch @jaycdave88! Could you please fill in the PR description with some context/motivation for the change? Thanks 🙏

@albertvaka no worries! I have updated the PR with some information. Please let me know if you need additional docs! Thanks

File: manifests/windows/agent6.pp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    package { 'Datadog Agent':
+      ensure            => installed,
+      provider          => 'windows',
+      source            => "$msi",
+      install_options   => ['/quiet', {'APIKEY' => $api_key, 'HOSTNAME' => $hostname, 'TAGS' => $tags}]
+    }
+
+    exec { 'DatadogRestart':
+      command           => "& 'C:\Program Files\Datadog\Datadog Agent\embedded\agent.exe' restart-service",
```
Comment (albertvaka) [inline manifests/windows/agent6.pp]: The CI seems to complain about these slashes being unescaped.

File: manifests/windows/agent6.pp (hunk 1 above)
Comment (jaycdave88) [reply manifests/windows/agent6.pp]: Ah, gotta love Sublime - ill fix that to all spaces, 2x.

File: manifests/windows/agent6.pp (hunk 2 above)
Comment (jaycdave88) [reply manifests/windows/agent6.pp]: yeah, I saw that ... I will try to switch the \ to / and see if that affects the agent install. I assume that will fix the CI.

File: manifests/windows/agent6.pp (hunk 1 above)
Comment (jaycdave88) [reply manifests/windows/agent6.pp]: Fixed.

**Comment by jaycdave88** [comment]
> Shouldn't we create a `datadog.yaml` on Windows as well? We could reuse the same code, and simply change the location for the file.
> 
> Otherwise (if we don't want to do this now), we should add a comment saying that config params set through puppet won't have an effect on Windows to avoid confusion.

@albertvaka - I have updated the logic to create a Windows Datadog config.

**Comment by albertvaka** [comment]
I'm trying to de-duplicate some code between Linux and Win. I have a question: Why is `$should_install_win` a parameter? In which case would you not want it to install the agent if it's not there and why only on Windows (since we didn't have such option on Linux)?

**Comment by jaycdave88** [comment]
> I'm trying to de-duplicate some code between Linux and Win. I have a question: Why is `$should_install_win` a parameter? In which case would you not want it to install the agent if it's not there and why only on Windows (since we didn't have such option on Linux)?

@albertvaka So that option was added to allow the user of the module to run a mass uninstall of the Datadog agent. The use cases I was trying to address, were Install and Uninstall - so I added that flag as a conditional to trigger a different workflow.

**Comment by albertvaka** [comment]
I've unified the two if branches for Linux and Windows into a single code path with some smaller ifs. I still would like to write some tests for the Windows code before we land this, but I think we are close. Thanks for your help :)

**Comment by jaycdave88** [comment]
> I've unified the two if branches for Linux and Windows into a single code path with some smaller ifs. I still would like to write some tests for the Windows code before we land this, but I think we are close. Thanks for your help :)

My pleasure! Please let me know if there is anything else that I can help out with! @albertvaka

**Comment by albertvaka** [comment]
I've moved this to a new PR: https://github.com/DataDog/puppet-datadog-agent/pull/557

**Pull request closed without merging**